src/device-state/db-format.ts

```diff
@@ -92,67 +88,77 @@ export async function setApps(
 /**
  * Create target state from database state
  */
-export async function getTargetJson(): Promise<TargetApps> {
+export async function getTargetWithRejections(): Promise<{
+	apps: TargetApps;
+	rejections: AppRelease[];
+}> {
 	const dbApps = await getDBEntry();
 
-	return dbApps
-		.map(
-			({
-				source,
-				uuid,
-				releaseId,
-				commit: releaseUuid,
-				...app
-			}): [string, TargetApp] => {
-				const services = (JSON.parse(app.services) as DatabaseService[])
-					.map(
-						({
-							serviceName,
-							serviceId,
-							imageId,
-							...service
-						}): [string, TargetService] => [
-							serviceName,
-							{
-								id: serviceId,
-								image_id: imageId,
-								..._.omit(service, ['appId', 'appUuid', 'commit', 'releaseId']),
-							} as TargetService,
-						],
-					)
-					// Map by serviceName
-					.reduce(
-						(svcs, [serviceName, s]) => ({
-							...svcs,
-							[serviceName]: s,
-						}),
-						{},
-					);
-
-				const releases = releaseUuid
-					? {
-							[releaseUuid]: {
-								id: releaseId,
-								services,
-								networks: JSON.parse(app.networks),
-								volumes: JSON.parse(app.volumes),
-							} as TargetRelease,
-						}
-					: {};
-
-				return [
-					uuid,
+	const apps: TargetApps = {};
+	const rejections: AppRelease[] = [];
+
+	for (const {
+		source,
+		rejected,
+		uuid,
+		releaseId,
+		commit: releaseUuid,
+		...app
+	} of dbApps) {
+		const services = Object.fromEntries(
+			(JSON.parse(app.services) as DatabaseService[]).map(
+				({
+					serviceName,
+					serviceId,
+					imageId,
+					// Ignore these fields
+					appId: _appId,
+					appUuid: _appUuid,
+					commit: _commit,
+					releaseId: _releaseId,
+					// Use the remainder of the fields for the service description
+					...service
+				}) => [
+					serviceName,
 					{
-						id: app.appId,
-						name: app.name,
-						class: app.class,
-						is_host: !!app.isHost,
-						releases,
-					},
-				];
-			},
-		)
-		.reduce((apps, [uuid, app]) => ({ ...apps, [uuid]: app }), {});
+						id: serviceId,
+						image_id: imageId,
+						...service,
+					} as TargetService,
+				],
+			),
+		);
+
+		const releases = releaseUuid
+			? {
+					[releaseUuid]: {
+						id: releaseId,
+						services,
+						networks: JSON.parse(app.networks),
+						volumes: JSON.parse(app.volumes),
+					} as TargetRelease,
```

Comment:
As mentioned elsewhere `satisfies` will increase the type safety
```suggestion
					} satisfies TargetRelease,
```
but in this case it's quite possible you'll want the
```suggestion
					} satisfies TargetRelease as TargetRelease,
```
version to keep the wider typing rather than the narrowed version that `satisfies` alone gives